framework/include/base/FEProblem.h

```diff
@@ -1210,4 +1224,43 @@ FEProblem::allowOutput(bool state)
   _app.getOutputWarehouse().allowOutput<T>(state);
 }
 
+
+template<typename T>
+void
+FEProblem::initializeUserObjects(const MooseObjectWarehouse<T> & warehouse)
+{
+  if (warehouse.hasActiveObjects())
+  {
+    for (THREAD_ID tid = 0; tid < libMesh::n_threads(); ++tid)
+    {
+      const std::vector<MooseSharedPointer<T> > & objects = warehouse.getActiveObjects(tid);
+      for (typename std::vector<MooseSharedPointer<T> >::const_iterator it = objects.begin(); it != objects.end(); ++it)
+        (*it)->initialize();
+    }
+  }
+}
+
+
+template<typename T>
+void
+FEProblem::finalizeUserObjects(const MooseObjectWarehouse<T> & warehouse)
+{
+  if (warehouse.hasActiveObjects())
+  {
+    const std::vector<MooseSharedPointer<T> > & objects = warehouse.getActiveObjects(0);
+    for (unsigned int i = 0; i < objects.size(); ++i)
```

Comment:
This shoulds be from thread 1 to _n_? Better yet, we should look into using parallel_reduce but I suppose that can happen later.